### ISSUE TITLE
loss become negative, when train the partical fc !

### ISSUE THREAD

**Issue by luckycaicai** [deepinsight/insightface#1428]
In the first stage of backward,
https://github.com/deepinsight/insightface/blob/48a9e0bb5268137ba8ba3391c94a7d4425ed75aa/recognition/arcface_torch/partial_fc.py#L153
To monitor changes of training loss, I use

`y = grad * logits`
`loss1 = y.sum()`
`loss1.backward()`

to replace it. 
No matter the nember of training data, after dozens of epochs, loss always becomes negative.
![image](https://user-images.githubusercontent.com/41863500/111112427-1e514600-859b-11eb-94ea-03d801efdd55.png)

**Comment by anxiangsir** [comment]
![image](https://user-images.githubusercontent.com/31175974/111139963-be6b9700-85bc-11eb-9940-f1110ff140e5.png)
Modifying the `backward of logits` means that the back propagation of the cross entropy will not be correct.

**Comment by luckycaicai** [comment]
The gradient of logits is not changed, like this in your code:


        ...

        ...
        logits = self.forward(total_features, norm_weight)
        logits = self.margin_softmax(logits, total_label)

        with torch.no_grad():
            max_fc = torch.max(logits, dim=1, keepdim=True)[0]
            dist.all_reduce(max_fc, dist.ReduceOp.MAX)

            # calculate exp(logits) and all-reduce
            logits_exp = torch.exp(logits - max_fc)
            logits_sum_exp = logits_exp.sum(dim=1, keepdims=True)
            dist.all_reduce(logits_sum_exp, dist.ReduceOp.SUM)

            # calculate prob
            logits_exp.div_(logits_sum_exp)

            # get one-hot
            grad = logits_exp
            index = torch.where(total_label != -1)[0]
            one_hot = torch.zeros(size=[index.size()[0], grad.size()[1]], device=grad.device)
            one_hot.scatter_(1, total_label[index, None], 1)

            # calculate loss
            loss = torch.zeros(grad.size()[0], 1, device=grad.device)
            loss[index] = grad[index].gather(1, total_label[index, None])
            dist.all_reduce(loss, dist.ReduceOp.SUM)
            loss_v = loss.clamp_min_(1e-30).log_().mean() * (-1)

            # calculate grad
            grad[index] -= one_hot
            grad.div_(self.batch_size * self.world_size)

Then  replace ' logits.backward(grad) ' by follow lines:

        y = grad * logits
        loss1 = y.sum()
        loss1.backward()

I think they are equivalent, so use the latter instead of "  logits.backward(grad) ". This replacement is just to print and monitor the changes of training loss.

**Comment by anxiangsir** [comment]
I think auto grad in this code doesn't work, because:
```python
 with torch.no_grad():
    pass
````
and a lot of communication function such as:
```python
dist.all_reduce
dist.all_gather
```
